File: Problems 0000-0099/23. Merge k Sorted Lists/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

import heapq
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        # return self.mergeKListsRange(lists, 0, len(lists) - 1)

        k = len(lists)
        h = []
        for i in range(k):
            if not lists[i]:
                continue
            heapq.heappush(h, (lists[i].val, i, lists[i]))
        
        dummy = curr = ListNode()
        while h:
            _, i, node = heapq.heappop(h)
            curr.next = node
            curr = curr.next
            if node.next:
                heapq.heappush(h, (node.next.val, i, node.next))
        
        return dummy.next

    # def mergeKListsRange(self, lists, start, end):
    #     if start > end:
    #         return
    #     if start == end:
    #         return lists[start]
        
    #     mid = start + (end - start) // 2
    #     left = self.mergeKListsRange(lists, start, mid)
    #     right = self.mergeKListsRange(lists, mid + 1, end)
    #     return self.mergeLists(left, right)

    # def mergeLists(self, l1, l2):
    #     dummy = ListNode()
    #     curr = dummy
    #     while l1 or l2:
    #         if l1 and l2:
    #             if l1.val <= l2.val:
    #                 curr.next = l1
    #                 l1 = l1.next
    #             else:
    #                 curr.next = l2
    #                 l2 = l2.next
    #         elif l1:
    #             curr.next = l1
    #             l1 = l1.next
    #         else:
    #             curr.next = l2
    #             l2 = l2.next
    #         curr = curr.next
        
    #     return dummy.next
        

File: Problems 0000-0099/23. Merge k Sorted Lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        return mergeKLists(lists, 0, lists.length - 1);
    }

    private ListNode mergeKLists(ListNode[] lists, int start, int end) {
        if (start > end) {
            return null;
        }
        if (start == end) {
            return lists[start];
        }
        int mid = start + (end - start) / 2;
        ListNode l1 = mergeKLists(lists, start, mid);
        ListNode l2 = mergeKLists(lists, mid + 1, end);

        return mergeLists(l1, l2);
    }

    public ListNode mergeLists(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        ListNode curr = dummy;

        while (l1 != null || l2 != null) {
            if (l2 == null || (l1 != null && l1.val < l2.val)) {
                curr.next = l1;
                l1 = l1.next;
            } else {
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }
        return dummy.next;
    }

    // public ListNode mergeKLists(ListNode[] lists) {
    //     PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a, b) -> a.val - b.val);

    //     for (ListNode node : lists) {
    //         if (node == null) {
    //             continue;
    //         }
    //         minHeap.add(node);
    //     }
        
    //     ListNode dummy = new ListNode();
    //     ListNode curr = dummy;
    //     while (!minHeap.isEmpty()) {
    //         ListNode node = minHeap.poll();
    //         curr.next = node;
    //         curr = curr.next;

    //         if (node.next != null) {
    //             minHeap.add(node.next);
    //         }
    //     }
    //     return dummy.next;
    // }
}